apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs

```diff
@@ -0,0 +1,129 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Cloud.Firestore.V1;
+using Google.Protobuf;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Cloud.Firestore.V1.StructuredAggregationQuery.Types;
+
+namespace Google.Cloud.Firestore;
+
+/// <summary>
+/// A query for running an aggregation over a [StructuredQuery][google.firestore.v1.StructuredQuery]. Currently only "count(*)" aggregation is supported.
+/// </summary>
+public sealed class AggregateQuery : IEquatable<AggregateQuery>
+{
+    private readonly Query _query;
+
+    /// <summary>
+    /// The "alias" to specify in the <see cref="RunAggregationQueryRequest"/> proto when running a count query.
+    /// The actual value is not meaningful, but will be used to get the count out of the <see cref="RunAggregationQueryResponse"/>.
+    /// </summary>
+    private const string CountAlias = "Count";
+
+    private readonly IReadOnlyList<Aggregation> _aggregations;
+
+    internal AggregateQuery(Query query) =>
+        _query = GaxPreconditions.CheckNotNull(query, nameof(query));
+
+    private AggregateQuery(Query query, IReadOnlyList<Aggregation> aggregations)
+    {
+        _query = GaxPreconditions.CheckNotNull(query, nameof(query));
+        _aggregations = aggregations;
+    }
+
+    internal AggregateQuery WithAggregation(Aggregation aggregation)
+    {
+        if (aggregation.Count != null)
+        {
+            aggregation.Alias = CountAlias;
+        }
+        var newAggregations = _aggregations == null ? new List<Aggregation>() : new List<Aggregation>(_aggregations);
+        newAggregations.Add(aggregation);
+        return new AggregateQuery(_query, newAggregations);
+    }
+
+    /// <summary>
+    /// Asynchronously takes a snapshot of the result after applying the aggregate functions on the query.
+    /// </summary>
+    /// <param name="cancellationToken">A cancellation token for the operation.</param>
+    /// <returns>A <see cref="AggregateQuerySnapshot"/> which contains results of the Aggregate functions.</returns>
+    public Task<AggregateQuerySnapshot> GetSnapshotAsync(CancellationToken cancellationToken = default) =>
+        GetSnapshotAsync(null, cancellationToken);
+
+    internal async Task<AggregateQuerySnapshot> GetSnapshotAsync(ByteString transactionId, CancellationToken cancellationToken)
+    {
+        IAsyncEnumerable<RunAggregationQueryResponse> responseStream = GetAggregationQueryResponseStreamAsync(transactionId, cancellationToken);
+        Timestamp? readTime = null;
+        Value count = null;
+        await responseStream.ForEachAsync(response =>
+        {
+            if (response.Result.AggregateFields != null)

```

Comment:
I'm not sure this is what you want - that means if there are multiple responses with `AggregateFields` populated, if the count is in the first one, we'll end up losing the count. You could do something like:

```csharp
if (count is null)
{
    response.Result.AggregateFields?.TryGetValue(CountAlias, out count);
}
```

Reply:
Actually, I'd also keep `count` as a `long?` variable, and have a check here - so something like:

```csharp
if (count is null && response.Result.AggregateFields?.TryGetValue(CountAlias, out var countValue) == true)
{
    // You probably need to change this - I can't remember exactly what it would be, but hopefully it'll be obvious enough in an IDE.
    GaxPreconditions.CheckState(countValue.ValueCase == Value.Types.ValueCase.Integer, "The count was not an integer.");
    count = countValue.IntegerValue;
}
```